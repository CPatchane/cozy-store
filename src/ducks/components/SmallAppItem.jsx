import React from 'react'

import Icon from 'cozy-ui/react/Icon'
import { translate } from 'cozy-ui/react/I18n'

import defaultAppIcon from '../../assets/icons/icon-cube.svg'
import { Link } from 'react-router-dom'
import Button from 'cozy-ui/react/Button'

export const SmallAppItem = ({
  t,
  slug,
  developer,
  editor,
  icon,
  name,
  namePrefix,
  installed,
  onClick,
  installedAppLink
}) => {
  const onShortcutAppButton = (e, link) => {
    // prevent from opening parent SmallAppItem link
    // which will lead to the application main page
    e.stopPropagation()
    if (link) window.location.assign(link)
  }
  return (
    // HACK a11y
    // `onKeyDown={(e) => e.keyCode === 13 ? onClick() : null`
    // This syntax make the div accessible.
    <div
      className='sto-small-app-item'
      onKeyDown={e => (e.keyCode === 13 ? onClick() : null)}
      onClick={onClick}
      tabIndex={0}
    >
      {icon ? (
        <img
          src={icon}
          alt={`${slug}-icon`}
          width='64'
          height='64'
          className='sto-small-app-item-icon'
        />
      ) : (
        <Icon
          className='sto-small-app-item-icon--default blurry'
          width='48'
          height='64'
          icon={defaultAppIcon}
          color='#95999D'
        />
      )}
      {installed && (
        <Icon
          className='sto-small-app-item-badge-installed'
          icon='check'
          color='#2bba40'
        />
      )}
      <div className='sto-small-app-item-desc'>
        <h4 className='sto-small-app-item-title'>
          {namePrefix && `${namePrefix} `}{name}
        </h4>
        <p className='sto-small-app-item-detail'>
          {developer.name === 'Cozy' ? 'Cozy Cloud Inc.' : developer.name}
        </p>
        {installed
          ? <Button
            onClick={(e) => onShortcutAppButton(e, installedAppLink)}
            label={t('app_item.open')}
            theme='secondary'
            className='sto-small-app-item-button'
            size='tiny'
          />
          : <Link
            to={`/discover/${slug}/manage`}
            onClick={(e) => onShortcutAppButton(e)}
            className='c-btn c-btn--regular c-btn--tiny sto-small-app-item-button sto-small-app-item-button-install'
          >
            <span>{t('app_item.install')}</span>
          </Link>
        }
      </div>
    </div>
  )
}

export default translate()(SmallAppItem)
