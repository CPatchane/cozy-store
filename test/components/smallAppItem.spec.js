'use strict'

/* eslint-env jest */

import React from 'react'
import { shallow } from 'enzyme'

import SmallAppItem from '../../src/ducks/components/SmallAppItem'

const appMock = {
  slug: 'test',
  editor: 'cozy',
  developer: {
    name: 'Cozy'
  },
  icon: '<svg></svg>',
  name: 'Test',
  version: '3.0.3',
  onClick: jest.fn()
}

const appMock2 = {
  slug: 'test2',
  editor: '',
  developer: {
    name: 'Naming me'
  },
  icon: '<svg></svg>',
  name: 'Test2',
  version: '3.0.3-beta7483',
  installed: true,
  onClick: jest.fn()
}

const appMockWithoutIcon = {
  slug: 'test2',
  editor: '',
  developer: {
    name: 'Naming me'
  },
  icon: '',
  name: 'Test2',
  version: '3.0.3-beta7483',
  installed: true,
  onClick: jest.fn()
}

describe('SmallAppItem component', () => {
  it('should be rendered correctly an app', () => {
    const component = shallow(
      <SmallAppItem {...appMock} />
    ).node
    expect(component).toMatchSnapshot()
  })

  it('should be rendered correctly an installed app', () => {
    const component = shallow(
      <SmallAppItem {...appMock2} />
    ).node
    expect(component).toMatchSnapshot()
  })

  it('should be rendered correctly an installed app', () => {
    const component = shallow(
      <SmallAppItem {...appMockWithoutIcon} />
    ).node
    expect(component).toMatchSnapshot()
  })

  it('should not call onClick when Enter is not pressed', () => {
    const component = shallow(
      <SmallAppItem {...appMock} />
    )
    expect(component.find('div.sto-small-app-item').length).toBe(1)
    component.find('div.sto-small-app-item').simulate('keydown', {keyCode: 98})
    expect(appMock.onClick.mock.calls.length).toBe(0)
  })

  it('should call onClick when Enter is pressed', () => {
    const component = shallow(
      <SmallAppItem {...appMock} />
    )
    expect(component.find('div.sto-small-app-item').length).toBe(1)
    component.find('div.sto-small-app-item').simulate('keydown', {keyCode: 13})
    expect(appMock.onClick.mock.calls.length).toBe(1)
  })
})
