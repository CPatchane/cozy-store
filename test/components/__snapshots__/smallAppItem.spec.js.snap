// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SmallAppItem component should be rendered correctly an app 1`] = `
<div
  className="sto-small-app-item"
  onClick={[MockFunction]}
  onKeyDown={[Function]}
  tabIndex={0}
>
  <img
    alt="test-icon"
    className="sto-small-app-item-icon"
    height="64"
    src="<svg></svg>"
    width="64"
  />
  <div
    className="sto-small-app-item-desc"
  >
    <div
      className="sto-small-app-item-text"
    >
      <h4
        className="sto-small-app-item-title"
      >
        Cozy Test
      </h4>
      <p
        className="sto-small-app-item-detail"
      >
        Cozy Cloud Inc.
      </p>
    </div>
    <div
      className="sto-small-app-item-buttons"
    >
      <Link
        className="c-btn c-btn--regular c-btn--tiny sto-small-app-item-button sto-small-app-item-button-install"
        onClick={[Function]}
        replace={false}
        to="/discover/test/manage"
      >
        <span>
          Install
        </span>
      </Link>
    </div>
  </div>
</div>
`;

exports[`SmallAppItem component should be rendered correctly an installed app 1`] = `
<div
  className="sto-small-app-item"
  onClick={[MockFunction]}
  onKeyDown={[Function]}
  tabIndex={0}
>
  <img
    alt="test2-icon"
    className="sto-small-app-item-icon"
    height="64"
    src="<svg></svg>"
    width="64"
  />
  <div
    className="sto-small-app-item-desc"
  >
    <div
      className="sto-small-app-item-text"
    >
      <h4
        className="sto-small-app-item-title"
      >
        Test2
      </h4>
      <p
        className="sto-small-app-item-detail"
      >
        Naming me
      </p>
    </div>
    <div
      className="sto-small-app-item-buttons"
    >
      <DefaultButton
        className="sto-small-app-item-button"
        label="Open"
        onClick={[Function]}
        size="tiny"
        theme="secondary"
      />
    </div>
  </div>
</div>
`;

exports[`SmallAppItem component should be rendered correctly an installed app without icon provided 1`] = `
<div
  className="sto-small-app-item"
  onClick={[MockFunction]}
  onKeyDown={[Function]}
  tabIndex={0}
>
  <Icon
    className="sto-small-app-item-icon--default blurry"
    color="#95999D"
    height="64"
    icon="test-file-stub"
    spin={false}
    width="48"
  />
  <div
    className="sto-small-app-item-desc"
  >
    <div
      className="sto-small-app-item-text"
    >
      <h4
        className="sto-small-app-item-title"
      >
        Cozy Test2
      </h4>
      <p
        className="sto-small-app-item-detail"
      >
        Naming me
      </p>
    </div>
    <div
      className="sto-small-app-item-buttons"
    >
      <DefaultButton
        className="sto-small-app-item-button"
        label="Open"
        onClick={[Function]}
        size="tiny"
        theme="secondary"
      />
    </div>
  </div>
</div>
`;
